File: samples/pascal/csvstats.pas
{ Summary statistics for a column of a comma-separated file. }
unit CsvStats;

{$MODE OBJFPC}{$H+}

interface

uses
  SysUtils, Classes;

type
  { One reading, as it comes out of the file. }
  TSample = record
    Name: string;
    Value: Double;
  end;

  ESampleError = class(Exception);

  { A named column of numbers. }
  TColumn = class(TObject)
  private
    FName: string;
    FValues: array of Double;
    function GetCount: Integer;
  public
    constructor Create(const AName: string);
    destructor Destroy; override;
    procedure Add(const Value: Double);
    function Mean: Double;
    function Deviation: Double;
    property Name: string read FName;
    property Count: Integer read GetCount;
  end;

function Largest(const Values: array of Double): Double;

{ Promised here, and never written below: the unit will not link until
  somebody writes it. It is left that way on purpose. }
procedure WriteReport(const Path: string);

implementation

uses
  Math;

constructor TColumn.Create(const AName: string);
begin
  inherited Create;
  FName := AName;
  SetLength(FValues, 0);
end;

destructor TColumn.Destroy;
begin
  SetLength(FValues, 0);
  inherited Destroy;
end;

function TColumn.GetCount: Integer;
begin
  Result := Length(FValues);
end;

procedure TColumn.Add(const Value: Double);
begin
  SetLength(FValues, Length(FValues) + 1);
  FValues[High(FValues)] := Value;
end;

function TColumn.Mean: Double;
var
  Value: Double;
  Total: Double;
begin
  if Length(FValues) = 0 then
    raise ESampleError.CreateFmt('column %s is empty', [FName]);
  Total := 0;
  for Value in FValues do
    Total := Total + Value;
  Result := Total / Length(FValues);
end;

function TColumn.Deviation: Double;
var
  Value: Double;
  Average: Double;
  Total: Double;
begin
  Average := Mean;
  Total := 0;
  for Value in FValues do
    Total := Total + Sqr(Value - Average);
  Result := Sqrt(Total / Length(FValues));
end;

function Largest(const Values: array of Double): Double;
var
  Value: Double;
begin
  Result := Values[Low(Values)];
  for Value in Values do
    if Value > Result then
      Result := Value;
end;

end.

File: samples/lua/src/csvstats.lua
--- Summary statistics for a column of a comma-separated file.
-- @module csvstats

local io = require("io")
local math = require("math")

local M = {}

local Column = {}
Column.__index = Column

--[[
An earlier version kept a running median here. It was wrong for an even
number of samples, and nothing used it:

function Column:median()
    return self.values[#self.values / 2]
end
]]

--- Makes an empty column.
-- @param name the column's heading
function Column.new(name)
    local self = setmetatable({}, Column)
    self.name = name
    self.values = {}
    return self
end

function Column:add(value)
    self.values[#self.values + 1] = value
end

function Column:mean()
    if #self.values == 0 then
        return nil
    end
    local total = 0
    for _, value in ipairs(self.values) do
        total = total + value
    end
    return total / #self.values
end

function Column:deviation()
    local mean = self:mean()
    if mean == nil then
        return nil
    end
    local sum = 0
    for _, value in ipairs(self.values) do
        sum = sum + (value - mean) ^ 2
    end
    return math.sqrt(sum / #self.values)
end

local function parse_line(line)
    local fields = {}
    for field in line:gmatch("[^,]+") do
        fields[#fields + 1] = field
    end
    return fields
end

--- Reads a file and returns one Column per heading.
function M.read(path)
    local handle = assert(io.open(path, "r"))
    local headings = parse_line(handle:read("l"))
    local columns = {}
    for _, heading in ipairs(headings) do
        columns[#columns + 1] = Column.new(heading)
    end
    for line in handle:lines() do
        local fields = parse_line(line)
        for index, field in ipairs(fields) do
            local number = tonumber(field)
            if number ~= nil then
                columns[index]:add(number)
            end
        end
    end
    handle:close()
    return columns
end

M.Column = Column

return M
